### DISCUSSION TITLE
LowLevelApi/Semantic/Example

### DISCUSSION THREAD

**Discussion by dduquesne** [scala/scala3#12575] [Metaprogramming]
Hello,

I have some issue understanding the low level API.
Especially these 4 objects below :

+- Term --------+- Ref -+- Ident
                                      +- Select

+- TypeRepr -+- NamedType -+- TermRef
                                               +- TypeRef

I have found few information/example in Scala3 syntaxe summary.
But i am still lost and do not see completely the source code represented by each one.
Any example would be realy welcome.

Thank for your help,

**Comment by nicolasstucki** [comment]
```scala
trait A { type X; val x: X }
val a: A = ...
val b: a.X = a.x
val c: a.type = a
```

In the example above `a` will be an `Ident` and `a.x` will be a `Select`. Selects usually appear when there is a prefix before term that you are referring to. Many times a `Select` can be represented as an equivalent `Ident` but the `Select` has more information about positions and what the user really wrote.

The `A` and `a.X` are `TypeRef` as they refer to a type. The `a` part of  `a.X` and ` a.this` are `TermRef` as these are parts of the type that refer to a term.
